Auth0/OAuth2Grant.swift

```diff
@@ -67,50 +81,99 @@ struct PKCE: OAuth2Grant {
+                  maxAge: maxAge)
     }
 
-    init(authentication: Authentication, redirectURL: URL, verifier: String, challenge: String, method: String, responseType: [ResponseType], nonce: String? = nil) {
+    init(authentication: Authentication,
+         redirectURL: URL,
+         verifier: String,
+         challenge: String,
+         method: String,
+         responseType: [ResponseType],
+         nonce: String? = nil,
+         leeway: Int,
+         maxAge: Int? = nil) {
         self.authentication = authentication
         self.redirectURL = redirectURL
         self.verifier = verifier
         self.responseType = responseType
-
+        self.leeway = leeway
+        self.maxAge = maxAge
         var newDefaults: [String: String] = [
             "code_challenge": challenge,
             "code_challenge_method": method
         ]
-
         if let nonce = nonce {
             newDefaults["nonce"] = nonce
         }
-
         self.defaults = newDefaults
     }
 
     func credentials(from values: [String: String], callback: @escaping (Result<Credentials>) -> Void) {
-        guard
-            let code = values["code"]
-            else {
-                let string = "No code found in parameters \(values)"
-                return callback(.failure(error: AuthenticationError(string: string)))
+        guard let code = values["code"] else {
+            let string = "No code found in parameters \(values)"
+            return callback(.failure(error: AuthenticationError(string: string)))
         }
-        guard validate(responseType: self.responseType, token: values["id_token"], nonce: self.defaults["nonce"]) else {
-            return callback(.failure(error: WebAuthError.invalidIdTokenNonce))
-        }
-        let clientId = self.authentication.clientId
-        self.authentication
-            .tokenExchange(withCode: code, codeVerifier: verifier, redirectURI: redirectURL.absoluteString)
-            .start { result in
-                // Special case for PKCE when the correct method for token endpoint authentication is not set (it should be None)
-                if case .failure(let cause as AuthenticationError) = result, cause.description == "Unauthorized" {
-                    let error = WebAuthError.pkceNotAllowed("Unable to complete authentication with PKCE. PKCE support can be enabled by setting Application Type to 'Native' and Token Endpoint Authentication Method to 'None' for this app at 'https://manage.auth0.com/#/applications/\(clientId)/settings'.")
-                    callback(Result.failure(error: error))
-                } else {
-                    callback(result)
-                }
+        let idToken = values["id_token"]
+        let responseType = self.responseType
+        let authentication = self.authentication
+        let leeway = self.leeway
+        let nonce = self.defaults["nonce"]
+        let maxAge = self.maxAge
+        let isHybridFlow = responseType.contains(.idToken)
```

Comment:
hybrid flow examples:
- token id_token code
- token code
- id_token code

You could rename to `frontChannelIdTokenExpected`

Reply:
Changed in ca28451be12118c8928745aead45feaa63e669e1.